### PR TITLE
automating analytics

### PR DESCRIPTION
**Pull request by SungjunApp**
ticket: https://pixlee.atlassian.net/browse/EN-3568

# Supported automatic Analytics events

### PXLAlbum
- Loadmore: when you use `PXLClient.sharedClient.loadNextPageOfPhotosForAlbum(album: album)` and load the second or the next pages, we fire `loadmore` events for you.

## UI components
UI components now have an option of firing analytics events automatically. 

### PXLGridView (similar to Widget)
- VisibleWidget
- OpenedWidget

###  PXLPhotoProductView (similar to Lightbox)
- OpenLightbox

# How to enable it?
- you need to set `PXLClient.sharedClient.autoAnalyticsEnabled = true` only once. Then, the SDK manages the analytics events explained above for you except PXLGridView. For PXLGridView, you need a little bit of implementation to pass your pagination information and widget type to PXLGridView. 
- If you set `PXLClient.sharedClient.autoAnalyticsEnabled = false`, which is the default, you can still manually fire the events, the same as before.

###  PXLGridView
```swift
#!swift
override func viewDidLoad() {
    PXLClient.sharedClient.apiKey = your api key
    PXLClient.sharedClient.secretKey = your secret key
    PXLClient.sharedClient.autoAnalyticsEnabled = true // <----- This activates auto-analytics
    PXLClient.sharedClient.regionId = your region id // <--- set it if you use multi-region.

    var gridView = PXLGridView()
    ...
    gridView.autoAnalyticsDelegate = self // <-- MUST be implemented
    ....
}

// this must be implemented to use this feature
extension AutoUIImageListViewController: PXLGridViewAutoAnalyticsDelegate {
    func setupAlbumForAutoAnalytics() -> (album: PXLAlbum, widgetType: String) {
        (album, "customized_widget_type")
    }
}

```
###  PXLPhotoProductView
```swift
#!swift

PXLClient.sharedClient.apiKey = your api key
PXLClient.sharedClient.secretKey = your secret key
PXLClient.sharedClient.autoAnalyticsEnabled = true // <----- This activates auto-analytics
PXLClient.sharedClient.regionId = your region id // <--- set it if you use multi-region.

let widget = PXLPhotoProductView.widgetForPhoto(photo: photo, delegate: self)
widget.setContent(...)
...
```

# UI Tests are added
- This is to monitor whether analytics events are successfully fired on UI elements.
- UI tests run on Circle CI. 

### How the tests work for analytics (Similar to how the Android SDK's testing works)
1. The tests mimic user behaviors by clicking and swiping views on the UI Components to fire an account's real analytics. 
2. Succeeded and failed analytics events will be broadcasted within the app via NotificationCenter. 
3. Then, tests can listen to capture the notification events to make sure that they are fired.
4. If everything works fine on all steps above, the tests passed.

### CODE REVIEW

**Review by jchen123 — APPROVED** [review]
this is so awesome. my question is for current users who may or may not be firing their events manually, if they upgraded to this, then their analytics should not change at all right?  its only when they manually toggle it to be automatic, then remove their manual event firing, then it will be automatic?  it would be bad if they upgraded and then started seeing double events

File: Classes/SDK/Widgets/PXLGridView.swift [hunk 1]
```diff
@@ -220,6 +228,91 @@ public class PXLGridView: UIView {
...
+            let viewFrame = inView.convert(view.bounds, from: view)
+            if viewFrame.intersects(inView.bounds) {
+                return isVisible(view: view, inView: inView.superview)
+            }
+            return false
+        }
+        return isVisible(view: view, inView: view.superview)
+    }
+    
+    func fireAnalytics() {
```
Comment (jchen123) [inline Classes/SDK/Widgets/PXLGridView.swift:270]: i think this shoul be a better name than just fireAnalytics. even if it were more verbose like fireOpenAndVisible

File: Classes/SDK/Widgets/PXLGridView.swift [hunk 2]
```diff
@@ -80,12 +84,12 @@ public class PXLGridView: UIView {
             Timer.scheduledTimer(withTimeInterval: 0.3, repeats: false) { _ in
                 self.logFirstHighlights()
             }
+            fireAnalytics()
```
Comment (jchen123) [inline Classes/SDK/Widgets/PXLGridView.swift:87]: does this fire without checking for a boolean for automatic analytics on or off?

**Comment by SungjunApp** [comment]
@jchen123 to use this feature, they have to explicitly enable the auto-analytics feature on each UI component by using <view>.enableAutoAnalytics(...) except `loadmore`.

'loadmore' has become to be always on so can have double events based on this structure. I'll change this to be off by default and add an argument to turn it on.

File: Classes/SDK/Widgets/PXLGridView.swift (hunk 2 above)
Comment (SungjunApp) [reply Classes/SDK/Widgets/PXLGridView.swift:87]: There are codes to check this.

they are enabled via <instance of PXLGridView>.enableAutoAnalytics(album: PXLAlbum, widgetType: String). If both album: PXLAlbum and widgetType: String are not nil (not null), they are activated. 

They are varified in fireAnalytics() -> [fireAnalyticsOpenedWidget() / fireAnalyticsWidgetVisible()].

**Review by jchen123 — CHANGES_REQUESTED** [review]
switching to "request changes" given our discussion for this and Android

File: Classes/SDK/Widgets/PXLGridView.swift [hunk 3]
```diff
@@ -220,6 +240,78 @@ public class PXLGridView: UIView {
...
+        }
+    }
+    
+    private func getAutoAnalytics() -> (album: PXLAlbum, widgetType: String)? {
+        if !PXLClient.sharedClient.autoAnalyticsEnabled {
+            return nil
+        }
+        
+        if autoAnalyticsDelegate == nil {
+            print("autoAnalyticsDelegate or PXLGridViewDelegate.setupAlbumForAutoAnalytics(){..} is not spcified")
```
Comment (jchen123) [inline Classes/SDK/Widgets/PXLGridView.swift:307]: specified spelling

File: Classes/SDK/Widgets/PXLGridView.swift [hunk 4]
```diff
@@ -220,6 +240,78 @@ public class PXLGridView: UIView {
...
+        } else {
+            print("You need to spcify 'class YourViewController PXLGridViewDelegate.setupAlbumForAutoAnalytics(){return (yourAlbum, your widget type)}'")
```
Comment (jchen123) [inline Classes/SDK/Widgets/PXLGridView.swift:309]: specify spelling

**Review by jchen123 — APPROVED** [review]
it all looks right. i'd feel comfortable if we manually tested that events dont get fired when auto analytics is off, but seems good

**Comment by SungjunApp** [comment]
I manually tested analytics by turning it off and on, testing the views, and iterating it. also added a UI test to verify this. All look fine.

**Pull request merged**